docs/user_guide/adversarial.rst

```diff
@@ -0,0 +1,281 @@
+For details, see [#4]_.
+
+Firstly, we will dicuss the models that this implementation accepts in
+:ref:`models`.
+Secondly, in :ref:`data_preprocessing`, we discuss the required
+data preprocessing, as :math:`X` must be an array of
+floats. Then, in :ref:`loss_functions`,
+we describe what are good choices of :math:`L_P` and :math:`L_A`.
+Lastly, in :ref:`training` we give some
+useful tips to keep in mind when training this model, as
+adversarial methods such as these
+are inherently difficult to train.
+
+.. _models:
+
+Models
+~~~~~~
+
+One can implement the predictor and adversarial neural networks as
+a `torch.nn.Module` (using PyTorch) or as a `tensorflow.keras.Model` (using TensorFlow).
+This implementation has a soft dependency on either PyTorch or TensorFlow, and the user
+needs to have installed either one of the two soft dependencies. It is not possible to
+mix these dependencies, so a PyTorch predictor with TensorFlow loss functions are not
+possible.
+
+It is very important to define the neural network models with no activation function
+or discrete prediction function on the final layer. So, for instance, when predicting
+a categorical feature that is one-hot-encoded, the neural network should output the
+logits, not the one-hot-encoded discrete prediction.
+
+.. testcode::
+
+    predictor_model = tf.keras.Sequential([
+        tf.keras.layers.Dense(50, activation='relu'),
+        tf.keras.layers.Dense(1)
+    ])
+    adversary_model = tf.keras.Sequential([
+        tf.keras.layers.Dense(3, activation='relu'),
+        tf.keras.layers.Dense(1)
+    ])
+    mitigator = AdversarialFairness(
+        predictor_model=predictor_model,
+        adversary_model=adversary_model
+    )
+
+For simple or exploratory use cases, Fairlearn provides a very basic neural
+network builder.
+Instead of a neural network model, we can pass a list of keywords
+:math:`k_1, k_2, \dots` that indicate either
+the number of nodes :math:`k_i` (if :math:`k_i` is integer) or a keyword
+that indicates an activation function (if :math:`k_i` is a string) or
+a layer or activation function instance directly (if :math:`k_i` is callable).
+However, the number of nodes in the input
+and output layer are automatically inferred from data, and the final activation
+function (such as softmax for categorical predictors) are inferred from data.
+So, in the following example, the predictor model is
+a neural network with an input layer of
+the appropriate number of nodes, a hidden layer with 50 nodes and
+ReLU activations, and an output layer with an appropriate activation function.
+
+.. testcode::
+
+    mitigator = AdversarialFairness(
+        predictor_model=[50, "relu"],
+        adversary_model=[3, "relu"]
+    )
+
+Though, we
+do recommend you create your own neural networks beforehand,
+as you can tweak these to your
+liking.
+
+.. _data_preprocessing:
+
+Data Preprocessing
+~~~~~~~~~~~~~~~~~~
+
+In contrast to other mitigation techniques, we require you to
```

Comment:
Stylistic note.... I'm seeing "the user" rather than "you" in other places.

Reply:
Oh that is good to know, I'll adjust this, thanks